File: base/src/main/java/com/lengjiye/base/application/MasterApplication.java
package com.lengjiye.base.application;

import android.content.Context;

import com.lengjiye.base.inter.IApp;

public class MasterApplication implements IApp {

    private IApp iApp;

    @Override
    public Context applicationContext() {
        if (iApp == null) {
            return null;
        }
        return iApp.applicationContext();
    }

    @Override
    public String applicationId() {
        if (iApp == null) {
            return null;
        }
        return iApp.applicationId();
    }

    @Override
    public int versionCode() {
        if (iApp == null) {
            return 0;
        }
        return iApp.versionCode();
    }

    @Override
    public String versionName() {
        if (iApp == null) {
            return null;
        }
        return iApp.versionName();
    }

    @Override
    public Boolean isDebug() {
        if (iApp == null) {
            return null;
        }
        return iApp.isDebug();
    }

    @Override
    public String buildType() {
        if (iApp == null) {
            return null;
        }
        return iApp.buildType();
    }

    @Override
    public String baseUrl() {
        if (iApp == null) {
            return null;
        }
        return iApp.baseUrl();
    }

    @Override
    public void logout() {
        if (iApp == null) {
            return;
        }
        iApp.logout();
    }

    public void setIApp(IApp iApp) {
        this.iApp = iApp;
    }

    public static MasterApplication getInstance() {
        return Instance.masterApplication;
    }

    private static class Instance {
        private static MasterApplication masterApplication = new MasterApplication();
    }

}

File: base/src/main/java/com/lengjiye/base/application/MasterApplication.kt
package com.lengjiye.base.application

import android.content.Context
import com.lengjiye.base.inter.IApp

/**
 * 在 model 中使用 application
 */
class MasterApplication : IApp {

    private lateinit var iApp: IApp

    // get() 不能去掉 不然不能赋值
    override val applicationId: String?
        get() = iApp.applicationId

    override val application: Context?
        get() = iApp.application

    override val versionCode: Int?
        get() = iApp.versionCode

    override val versionName: String?
        get() = iApp.versionName

    override val isDebug: Boolean?
        get() = iApp.isDebug

    override val buildType: String?
        get() = iApp.buildType

    fun setIApp(iApp: IApp) {
        this.iApp = iApp
    }

    /**
     * 静态内部单例模式
     */
    private object Instance {
        val master = MasterApplication()
    }

    companion object {
        val instance = Instance.master
    }
}